src/rpc/README.txt

```diff
@@ -0,0 +1,60 @@
+--------------------------------------------------------------------------------
+Python -> OCaml RPC
+--------------------------------------------------------------------------------
+
+The mechanism implemented here and in `ocaml.py` in pysemgrep allows us to
+incrementally migrate functionality from Python to OCaml in support of our
+osemgrep effort. Instead of being limited to the one main call out to
+semgrep-core, we can now make additional calls from arbitrary points in the
+Python code.
+
+To add a function:
+
+- Update the `function_call` and `function_return` types in
+  `semgrep_output_v1.atd` in the `semgrep-interfaces` module.
+- Update `RPC.ml` to handle the function call and return the appropriate value.
+- Update `ocaml.py` to make the function call.
+- Call the function in `ocaml.py` from wherever you need to in the Python code.
+
+Suggested best practices:
+
+- Avoid exposing the ATD-generated types to clients of `ocaml.py`.
+- Do not perform any business logic in `ocaml.py`. It should just construct the
+  function call object, send it, receive the result, and return it.
+
+Invariants/limitations:
+
+- Function calls cannot be interleaved. The response to one call must be
+  received before a second call is sent.
+- We pay the performance cost of serializing and deserializing each argument and
+  return value. That cost is negligible for some workloads and is likely
+  unacceptable for others. Carefully consider how best to use this.
+- Further, the semgrep-core rpc process handles only a single call and then
+  dies. We could change that, but there would be a little bit more complexity
+  involved with managing a long-lived child process. Starting a process each
+  time adds some overhead, effectively preventing us from using this for
+  functionality that requires numerous function calls from Python to OCaml.
+
+PRC protocol details:
+
+- [Python] Serialize the function call to a string using ATD.
+- [Python] Send the string over the output channel:
+  - Compute the size of the string in bytes.
+  - Convert the size integer to a string, and terminate it with a newline.
```

Comment:
does it mean if the packet contains newlines we have a problem?

Reply:
Nope! I explicitly designed it to be robust in that case. Once we know the size from reading the header, we read `size` bytes from the stream without caring what their contents are. They can have newlines.